Fix bad popups offset in editor with single window off

jmb462 opened the pull request:
Fix #47217

### Issue description

When single window is disabled, popup menus in editor pops at a bad offset (see GIFs below)

### Fix proposal (Updated after Kobewi comment)

Using ``get_screen_transform()`` instead of ``get_global_transform()`` to set the popup position to deal with offset in multiwindow mode correctly.

### BEFORE
![bug_bezier](https://user-images.githubusercontent.com/3649998/131512600-caa65c0e-f603-4511-94c7-3b2abe16472f.gif)
Bezier Editor bug

![bug_script_editor](https://user-images.githubusercontent.com/3649998/131513032-636e3de0-07ad-4815-b145-018ced593789.gif)
Script Editor bug

![bug_shader](https://user-images.githubusercontent.com/3649998/131513057-8b4ca472-2017-48ed-a580-0934056dcc31.gif)
Shader Editor Bug

![bug_curve](https://user-images.githubusercontent.com/3649998/131512751-535a1f05-b693-49b4-b18e-afea0c4e8330.gif)
Curve editor bug

![bug_node2](https://user-images.githubusercontent.com/3649998/131513152-a7d47350-359b-4c00-a055-2e8890e1cde1.gif)
Canvas editor bug

### AFTER

![fix_bezier](https://user-images.githubusercontent.com/3649998/131513398-b0f5ccf7-0620-4026-8d3c-3fb5161d41a8.gif)
Bezier editor fixed

![fix_script_editor](https://user-images.githubusercontent.com/3649998/131513420-4d910b72-9865-436c-9315-a22ef276aa2e.gif)
Script editor fixed

![fix_script_shader](https://user-images.githubusercontent.com/3649998/131513481-d41c610c-73bc-4b43-8a94-e3ca874c54b4.gif)
Shader Editor Fixed

![fix_curve](https://user-images.githubusercontent.com/3649998/131513499-32d0e13d-7223-4218-a8a7-02d97c4e848b.gif)
Curve Editor Fixed

![fix_node2](https://user-images.githubusercontent.com/3649998/131513515-7d0c667c-d243-4e98-9b84-34c5b55f7a0f.gif)
Canvas Editor Fixed

KoBeWi commented:
This won't work if the control is rotated or scaled. It's not a problem in the editor, but e.g. TextEdit can't use this.

jmb462 committed:
Fix bad popups offset in editor with single window off

Co-authored-by: Gil Arasa Verge <gilarasaverge@gmail.com>
Co-authored-by: Tomasz Chabora <kobewi4e@gmail.com>

jmb462 commented:
> This won't work if the control is rotated or scaled. It's not a problem in the editor, but e.g. TextEdit can't use this. The problem might lie in `get_screen_transform()` possibly returning wrong value.

get_screen_transform() is working well.

All buggy popup were using get_global_transform() instead.

I've modified the PR to replace these occurence.
A lot of other popups uses get_screen_transform() and works well.

KoBeWi commented:
The change is good, but it doesn't fix all popups. I'd take the chance to replace all wrong usage of global position/transform. Also there are cases like e.g. ConnectionsDock, where the menu isn't properly trimmed:
![image](https://user-images.githubusercontent.com/2223172/131590574-e307aa97-c80a-4880-a651-9412e0b6fcce.png)
Such popups need `popup->set_size(Vector2())` call.

KoBeWi commented on editor/editor_inspector.cpp:
```diff
@@ -786,7 +786,7 @@ void EditorProperty::gui_input(const Ref<InputEvent> &p_event) {
 	} else if (mb.is_valid() && mb->is_pressed() && mb->get_button_index() == MOUSE_BUTTON_RIGHT) {
 		_ensure_popup();
-		menu->set_position(get_screen_position() + get_local_mouse_position());
+		menu->set_position(get_screen_transform().xform(get_local_mouse_position()));
```
I think this one is ok. If something already uses screen position then it can be left like that.

Also for other changes, I'd do
`get_global_position()` -> `get_screen_position()`
`get_global_transform().xform()` -> `get_screen_transform().xform()`
i.e. if something uses position, just fix the position and if it uses transform, fix transform.

jmb462 replied:
Thanks. Done.

groud commented:
This PR looks quite similar to https://github.com/godotengine/godot/pull/37506, but I am not sure that on e is still being worked on. You might want to have a look to it.

jmb462 commented:
> This PR looks quite similar to #37506, but I am not sure that on e is still being worked on. You might want to have a look to it.

@groud All cases fixed by this PR are also fixed in #37506 with the same modifications and #37506 is more exhaustive.

I think we could close this one and "reactivate" the review process of #37506.

pycbouh commented:
@jmb462 If you are up for it, it would likely be preferable if you were to incorporate the changes from the other PR here, add that other contributor as co-author and see it through to the end. The other PR is a year old, they may not be ready to work on it yet. But you are here 🙃

jmb462 commented:
OK, I've included most of @giarve modification in this PR (from #37506).
I've also found one or two popups that still use get_global_transform() and fixed it.
I've added some default size to popup to ensure they are trimmed properly (each time I can't find set_size() for these popups)

BUT I didn't include @giarve modification for following files cause it wasn't clear to me  :
- scene/main/window.cpp  related to screen set current screen
- platform/linuxbsd/display_server_x11.cpp cause I don't understand this part

KoBeWi commented:
Needs rebase, also `set_size(Vector2(1, 1)` needs to be changed to `reset_size()`.
Is this finished aside from that?

jmb462 commented:
> Needs rebase, also `set_size(Vector2(1, 1)` needs to be changed to `reset_size()`. Is this finished aside from that?

@KoBeWi Done. Seems finished to me.

TokageItLab commented:
Does it fix to #50720? I guess that issue caused by same reason.

KoBeWi commented:
So I *accidentally* opened [an extremely similar PR](https://github.com/godotengine/godot/pull/55297) 3 days ago, which has most of the changes that this one has. I made sure to go over every single popup, so you can take a look if your PR isn't missing something I did. I know that at least my changes to `visual_shader_editor_plugin.cpp` are still relevant, so you could just copy them over here.

jmb462 commented:
> So I _accidentally_ opened [an extremely similar PR](https://github.com/godotengine/godot/pull/55297) 3 days ago, which has most of the changes that this one has. I made sure to go over every single popup, so you can take a look if your PR isn't missing something I did. I know that at least my changes to `visual_shader_editor_plugin.cpp` are still relevant, so you could just copy them over here.

Done.

akien-mga commented:
I just created a new conflict by merging a new PR, so needs a rebase.

akien-mga commented:
> I just created a new conflict by merging another PR, so needs a rebase.

I did it myself and force pushed the rebase so that we can get it merged.

akien-mga commented:
For the record there's a few more places with `get_global_transform().xform` which might be worth reviewing, if they haven't been already:
```
$ rg "get_global_transform\(\).xform"
scene/3d/soft_dynamic_body_3d.cpp
403:                    PhysicsServer3D::get_singleton()->soft_body_move_point(physics_rid, r[i].point_index, r[i].spatial_attachment->get_global_transform().xform(r[i].offset));

scene/3d/visual_instance_3d.cpp
36:     return get_global_transform().xform(get_aabb());

scene/2d/node_2d.cpp
362:    return get_global_transform().xform(p_local);

editor/animation_track_editor.cpp
5229:                   box_selecting_from = scroll->get_global_transform().xform(mb->get_position());
5274:           Vector2 to = scroll->get_global_transform().xform(mm->get_position());

scene/3d/node_3d.cpp
757:    return get_global_transform().xform(p_local);

editor/plugins/node_3d_editor_plugin.cpp
6528:                                   aabb = collision_shape->get_global_transform().xform(collision_shape->get_shape()->get_debug_mesh()->get_aabb());
6534:                                           aabb.merge_with(col_shape->get_global_transform().xform(col_shape->get_shape()->get_debug_mesh()->get_aabb()));
```

KoBeWi commented:
These are unrelated. The PR was only about popups. `get_global_transform()` is in window's local coordinates, but popups expect screen coordinates.

akien-mga commented:
Thanks!
